### DISCUSSION TITLE
Mixing down two generated streams

### DISCUSSION THREAD

**Discussion by ld21** [pschatzmann/arduino-audio-tools#1449] [Q&A] [answered]
I would like to mix down two generated streams with the GeneratorMixer class. I couldn't find an example, so I tried to go from [https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction](https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction) and [https://pschatzmann.github.io/arduino-audio-tools/group__generator.html](https://pschatzmann.github.io/arduino-audio-tools/group__generator.html). 

However, with the following code, I get the error «no matching function for call to 'audio_tools::StreamCopy::StreamCopy(audio_tools::I2SStream&, audio_tools::GeneratorMixer<short int>&)'». 

What do I do wrong? Thank you for pointing it out.

Dani

`/**
 * @file streams-generator-i2s.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-i2s/README.md 
 * @copyright GPLv3
 */
 
#include "AudioTools.h"

AudioInfo info(44100, 1, 16); // Unbedingt auf 1 setzen mit MAX 89357A, sonst gibt's ein Klickedicklick.
SquareWaveGenerator<int16_t> squareWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
SineWaveGenerator<int16_t> sineWave(32000);                    // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> square(squareWave);              // Stream generated from square wave
GeneratedSoundStream<int16_t> sine(sineWave);                  // Stream generated from sine wave
GeneratorMixer<int16_t> mixer;

I2SStream out; 
StreamCopy copier(out, mixer);                                 // should copy mixed sound into i2s

int button_a_pin = 16;
bool button_a_was_pressed = false;

int button_b_pin = 4;
bool button_b_was_pressed = false;

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // Define streams to be mixed together
  mixer.begin();
  mixer.add(square);  
  mixer.add(sine);
    
  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);

  // Standard-I2S-Pins stehen unter: https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioConfig.h
  config.pin_ws = 15;
  config.pin_bck = 14;
  config.pin_data = 25;

  config.copyFrom(info); 
  out.begin(config);

  pinMode(button_a_pin,INPUT_PULLUP);
  pinMode(button_b_pin,INPUT_PULLUP);
}

// Arduino loop - copy sound to out 
void loop() {
  // -------- BUTTON A ----------
  if(digitalRead(button_a_pin) == LOW){
    if(!button_a_was_pressed){
      squareWave.begin(info, N_E5);
    }
    button_a_was_pressed = true;
  }
  else{
    if(button_a_was_pressed){
      squareWave.end();
    }
    button_a_was_pressed = false;
  }

  // -------- BUTTON B ----------
  if(digitalRead(button_b_pin) == LOW){
    if(!button_b_was_pressed){
      sineWave.begin(info, N_A5);
    }
    button_b_was_pressed = true;
  }
  else{
    if(button_b_was_pressed){
      sineWave.end();
    }
    button_b_was_pressed = false;
  }
  
  copier.copy();
}`

**Comment by pschatzmann** [comment, marked as answer]
Why don't you just read the chapter about mixing in the Wiki?

**Reply by ld21** [reply]
Thank you for pointing that out. I didn't realise I had to user InputMixer instead of GeneratorMixer (where would I have to use GeneratorMixer?). Your example from the Wiki works like a charm.

However – I would like to start and stop the generated streams from button presses. With the updated code (see below) I don't get any output. I however get a sound, if I add 
`  sineWave.begin(info, N_B5);
  squareWave.begin(info, N_E4);` 
in setup().

How would I start the generators as a result to a button press? Thank you very much for helping me. 
Dani

`#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"

AudioInfo info(32000, 1, 16); // Unbedingt auf 1 setzen mit MAX 89357A, sonst gibt's ein Klickedicklick.
SquareWaveGenerator<int16_t> squareWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
SineWaveGenerator<int16_t> sineWave(32000);                    // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> square(squareWave);              // Stream generated from square wave
GeneratedSoundStream<int16_t> sine(sineWave);                  // Stream generated from sine wave
InputMixer<int16_t> mixer;

I2SStream out; 
StreamCopy copier(out, mixer);                                 // should copy mixed sound into i2s

int button_a_pin = 16;
bool button_a_was_pressed = false;

int button_b_pin = 4;
bool button_b_was_pressed = false;

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);

  // Standard-I2S-Pins stehen unter: https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioConfig.h
  config.pin_ws = 15;
  config.pin_bck = 14;
  config.pin_data = 25;

  config.copyFrom(info); 
  out.begin(config);

  // Define streams to be mixed together
  mixer.add(square);  
  mixer.add(sine);
  mixer.begin(info);
    
  pinMode(button_a_pin,INPUT_PULLUP);
  pinMode(button_b_pin,INPUT_PULLUP);
}

// Arduino loop - copy sound to out 
void loop() {
  // -------- BUTTON A ----------
  if(digitalRead(button_a_pin) == LOW){
    if(!button_a_was_pressed){
      squareWave.begin(info, N_E5);
    }
    button_a_was_pressed = true;
  }
  else{
    if(button_a_was_pressed){
      squareWave.end();
    }
    button_a_was_pressed = false;
  }

  // -------- BUTTON B ----------
  if(digitalRead(button_b_pin) == LOW){
    if(!button_b_was_pressed){
      sineWave.begin(info, N_A5);
    }
    button_b_was_pressed = true;
  }
  else{
    if(button_b_was_pressed){
      sineWave.end();
    }
    button_b_was_pressed = false;
  }
  
  copier.copy();
}`

**Comment by pschatzmann** [comment]
I suggest that you have a look at the different streams-synth examples in the [examples-audiokit](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit) folder. 

I wasn't aware that the InputMixer even existed and reviewing the logic of the implementation did not make any sense to me, so I needed to correct it: You would combine multiple generators into a combined generator and then wrap this into a GeneratedSoundStream.

**Reply by ld21** [reply]
Thank you for pointing out these examples. However, the three I saw (streams-synthbasic1-audiokit, streams-synthbasic2-audiokit and streams-synth-audiokit) only work with one generated sound.

When you say 
> You would combine multiple generators into a combined generator and then wrap this into a GeneratedSoundStream.

Is there an example around too where I could look and try to understand?
Thank you, Dani

**Reply by pschatzmann** [reply]
No, there is no example: but going back to your initial error message: you try to use the GeneratorMixer as copy source: but this is a Generator and not a Stream. You would need to use
```C++
   GeneratorMixer<int16_t> mixer_gen;  
   GeneratedSoundStream<int16_t> mixed_stream(mixer_gen); // Stream generated from GeneratorMixer
   StreamCopy copier(out, mixed_stream); // should copy mixed sound into i2s
```

The examples were addressing your question about how to start and stop a sound and that does not change if the result is going finally thru a mixer

**Reply by ld21** [reply]
Hi Phil and everybody

I have been trying to adapt your example «streams-synthbasic1-audiokit.ino» to my hardware (ESP32 WROOM32E with MAX 89357A and some buttons). However I get confused since the audioAction()s are part of the AudioBoardStream kit which I had to (?) replace by the I2SStream out (?) because of my hardware (is that even correct?). 

The sketch now compiles but gives no sound if I press the buttons. 

The log says nothing but 
`[I] StreamCopy.h : 143 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops`

Do you see what I do wrongly?
Thank you, Dani

My code:
`/**
 * @file streams-synthbasic-audiokit.ino
 * @author Phil Schatzmann
 * @brief see https://www.pschatzmann.ch/home/2021/12/17/ai-thinker-audio-kit-building-a-simple-synthesizer-with-the-audiotools-library/)
 * @copyright GPLv3
 * Basierend auf Kommentar von PSchatzmann unter https://github.com/pschatzmann/arduino-audio-tools/discussions/1449
 * Versuch: An meinen Mini-Shooter-Board angepasstes Beispiel mit Vorlage unter https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-synthbasic1-audiokit/streams-synthbasic1-audiokit.ino
 * (funktioniert noch nicht)
 */
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"

AudioInfo info(44100, 1, 16); // Unbedingt auf 1 setzen mit MAX 89357A, sonst gibt's ein Klickedicklick.
AudioBoardStream kit(AudioKitEs8388V1);
SineWaveGenerator<int16_t> sine;
GeneratedSoundStream<int16_t> in(sine); 
I2SStream out; 
StreamCopy copier(out, in); 

void actionKeyOn(bool active, int pin, void* ptr){
  int freq = *((float*)ptr);
  sine.setFrequency(freq);
  in.begin();
}

void actionKeyOff(bool active, int pin, void* ptr){
  in.end();
}

// We want to play some notes on the AudioKit keys 
void setupActions(){
  // assign buttons to notes
  auto act_low = AudioActions::ActiveLow;
  static float note[] = {N_C3, N_D3, N_E3, N_F3, N_G3, N_A3}; // frequencies
  kit.audioActions().add(17, actionKeyOn, actionKeyOff, act_low, &(note[0])); // C3 // getKey gibt ein GPIO zurück // 17 left
  kit.audioActions().add(23, actionKeyOn, actionKeyOff, act_low, &(note[1])); // D3 // getKey gibt ein GPIO zurück // 23 right
  kit.audioActions().add(19, actionKeyOn, actionKeyOff, act_low, &(note[2])); // E3 // getKey gibt ein GPIO zurück // 19 up
  kit.audioActions().add(18, actionKeyOn, actionKeyOff, act_low, &(note[3])); // F3 // getKey gibt ein GPIO zurück // 18 down
  kit.audioActions().add(16, actionKeyOn, actionKeyOff, act_low, &(note[4])); // G3 // getKey gibt ein GPIO zurück // 16 shoot left
  kit.audioActions().add( 4, actionKeyOn, actionKeyOff, act_low, &(note[5])); // A3 // getKey gibt ein GPIO zurück //  4 shoot right
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial,AudioLogger::Info);

  // Setup output
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.pin_ws = 15;
  cfg.pin_bck = 14;
  cfg.pin_data = 25;
//  kit.sd_active = false;
  cfg.copyFrom(info); 
  out.begin(cfg);
//  kit.setVolume(80);

  // Setup sound generation based on AudioKit settins
  in.begin(cfg);

  // activate keys
  setupActions();

}

// copy the data
void loop() {
  copier.copy();
  kit.processActions();
}`

**Comment by pschatzmann** [comment]
Actually the class you are looking for is [AudioActions](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_audio_actions.html)!
e.g. 
```C++
AudioActions actions;
actions.add(17, actionKeyOn, actionKeyOff, ActiveLow, &(note[0])); 
actions.processActions();

```

**Reply by ld21** [reply]
That works, thank you very much. I miss the kit.setVolume() too – can I use that also without the kit?

**Comment by pschatzmann** [comment]
There is a chapter about volume management in the Wiki

**Reply by ld21** [reply]
Thanks. 

Trying to adapt to my situation from [https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control](https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control) I added the following:
`I2SStream out; 
VolumeStream volume(out);
StreamCopy copier(volume, in); `

Unfortunately, trying in setup() to change the volume with 
`volume.setVolume(0.1);` 
doesn't make a difference. What do I implement wrongly?

Cheers, Dani

**Reply by pschatzmann** [reply]
I guess you forgot to call volume.begin()

**Reply by ld21** [reply]
Perfect, volume can now be changed, thanks!

**Reply by ld21** [reply]
I tried to go with your example under [https://www.pschatzmann.ch/home/2021/12/17/ai-thinker-audio-kit-building-a-simple-synthesizer-with-the-audiotools-library/](https://www.pschatzmann.ch/home/2021/12/17/ai-thinker-audio-kit-building-a-simple-synthesizer-with-the-audiotools-library/) and add an effect. 

However, when I try to compile the code (below), it says

`synth_example_with_effect:16:1: error: invalid use of template-name 'audio_tools::AudioEffects' without an argument list`

Is your example compatible with an older version of the library? How should I code it otherwise? Thank you for a hint. Dani

```
#include "AudioTools.h"

AudioInfo info(32000, 1, 16); // Unbedingt auf 1 setzen mit MAX 89357A, sonst gibt's ein Klickedicklick.
AudioActions actions;
SquareWaveGenerator<int16_t> square(32000);
AudioEffects effects(square);
ADSRGain adsr(0.001,0.001,0.5, 0.005);
GeneratedSoundStream<int16_t> in(effects); 

I2SStream out; 
VolumeStream volume(out);
StreamCopy copier(volume, in); 

void actionKeyOn(bool active, int pin, void* ptr){
  int freq = *((float*)ptr);
  square.setFrequency(freq);
  adsr.keyOn();
}

void actionKeyOff(bool active, int pin, void* ptr){
  adsr.keyOff();
}

// We want to play some notes on the AudioKit keys 
void setupActions(){
  // assign buttons to notes
  auto act_low = AudioActions::ActiveLow;
  static float note[] = {N_C3, N_D3, N_E3, N_F3, N_G3, N_A3}; // frequencies
  actions.add(17, actionKeyOn, actionKeyOff, act_low, &(note[0])); // C3 // getKey gibt ein GPIO zurück // 17 left
  actions.add(23, actionKeyOn, actionKeyOff, act_low, &(note[1])); // D3 // getKey gibt ein GPIO zurück // 23 right
  actions.add(19, actionKeyOn, actionKeyOff, act_low, &(note[2])); // E3 // getKey gibt ein GPIO zurück // 19 up
  actions.add(18, actionKeyOn, actionKeyOff, act_low, &(note[3])); // F3 // getKey gibt ein GPIO zurück // 18 down
  actions.add(16, actionKeyOn, actionKeyOff, act_low, &(note[4])); // G3 // getKey gibt ein GPIO zurück // 16 shoot left
  actions.add( 4, actionKeyOn, actionKeyOff, act_low, &(note[5])); // A3 // getKey gibt ein GPIO zurück //  4 shoot right
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial,AudioLogger::Info);

  // Setup output
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.pin_ws = 15;
  cfg.pin_bck = 14;
  cfg.pin_data = 25;
  cfg.copyFrom(info); 
  out.begin(cfg);

  volume.begin();
  volume.setVolume(0.5);

  in.begin(cfg);

  effects.addEffect(adsr);
  
  // activate keys
  setupActions();

}

// copy the data
void loop() {
  copier.copy();
  actions.processActions();
}
```

**Reply by pschatzmann** [reply]
Don't use old examples from my blog because things are changing and I am only updating the corresponding examples in the examples directory of my project to make sure that they compile.
